src/app/lifecycle.ts

```diff
@@ -30,48 +31,97 @@ export class AppStore {
   private readonly appStore = new Map<string, FirebaseApp>();
 
   public initializeApp(options?: AppOptions, appName: string = DEFAULT_APP_NAME): App {
+    validateAppNameFormat(appName);
+
+    let autoInit = false;
     if (typeof options === 'undefined') {
+      autoInit = true
       options = loadOptionsFromEnvVar();
       options.credential = getApplicationDefault();
     }
 
-    if (typeof appName !== 'string' || appName === '') {
-      throw new FirebaseAppError(
-        AppErrorCodes.INVALID_APP_NAME,
-        `Invalid Firebase app name "${appName}" provided. App name must be a non-empty string.`,
-      );
-    } else if (this.appStore.has(appName)) {
-      if (appName === DEFAULT_APP_NAME) {
+    // Check if an App already exists and, if so, ensure its AppOptions match
+    // those of this `initializeApp` request. 
+    if (this.appStore.has(appName)) {
+      const currentApp = this.appStore.get(appName)!;
+      // Ensure the autoInit state matches the existing app's. If not, throw.
+      if (currentApp.autoInit() !== autoInit) {
         throw new FirebaseAppError(
-          AppErrorCodes.DUPLICATE_APP,
-          'The default Firebase app already exists. This means you called initializeApp() ' +
-          'more than once without providing an app name as the second argument. In most cases ' +
-          'you only need to call initializeApp() once. But if you do want to initialize ' +
-          'multiple apps, pass a second argument to initializeApp() to give each app a unique ' +
-          'name.',
-        );
+          AppErrorCodes.INVALID_APP_OPTIONS,
+          `Firebase app named "${appName}" attempted mismatch between custom AppOptions` +
+          ' and an App created via Auto Init.'
+        )
+      } else if (autoInit) {
+        // Auto-initialization is triggered when no options were passed to
+        // initializeApp. With no options to compare, simply return the App.
+        return currentApp;
       } else {
-        throw new FirebaseAppError(
-          AppErrorCodes.DUPLICATE_APP,
-          `Firebase app named "${appName}" already exists. This means you called initializeApp() ` +
-          'more than once with the same app name as the second argument. Make sure you provide a ' +
-          'unique name every time you call initializeApp().',
-        );
+        // Auto-initialization was not used. Ensure the options don't contain
+        // incomparable fields, and that options matches the existing app's.
+
+        // `httpAgent` objects cannot be compared with deepEquals impls.
+        // Idempotency cannot be supported if one exists.
+        if (typeof options.httpAgent !== 'undefined') {
```

Comment:
nit: Would it make sense to move this (and below) validation to a separate function (out of `initializeApp`) to clean it up a bit?